### ISSUE TITLE
ColorMode() as global state

### ISSUE THREAD

**Issue by workergnome** [processing/p5.js#696]
As we've been digging through the `colorMode` codebase as part of our conversations about HSB/HSL/RGB et al, we've run into a conceptual problem.

A lot of this has to do with unintuitive behavior in colorMode.  Intuitively, I assume that the colorMode set is the mode that colors are used in.  For example:

```javascript

  colorMode(RGB, 255);
  var c = color(127,255,0);

  colorMode(RGB,1);
  var myColor = c.getRed();
  print(myColor);	

```

I would assume, because I have set the global `colorMode`, that myColor would be equal to ~0.5.  Of course, it's not; it's equal to 127 just as I set it.  This is because what `colorMode` does is set *the way that colors are created*, not *what mode colors are in*.

My intuition says that setting HSB/RGB this way made sense, because it's easier than `color(100,200,100,RGB)` or rgbColor(100,200,300)`, but once we add the ability to scale colors values as a global state, it causes complications.

For things like `fill()` or `stroke()`, global states make sense, because the things that use them are just side effects, not objects with persistent state.  `colorMode()` is different—it involves persistence, and that means that we need to think harder about what it actually means.

Having an object whose functionality depends on globally set states also seems dirty to me.

### Potential Options:

1. We could calculate the `red()`, `green()`, `blue()`, `brightness()`, `saturation()`, `hue()`, et al using the global color mode.  There would be a performance hit here, but maybe not a large one, and it would make the software more intuitive.
2. We could modify the documentation to better explain what `colorMode()` does.
3. We could reconsider colorMode() as a global flag
5. We could create explicit constructors for various modes and scalings. 
4. Someone could explain to me what I have missed, we could close this issue, and I could go back to sleeping nights.

**Comment by workergnome** [comment]
pinging @cambridgemike, @kadamwhite, @sarahgp

**Comment by sarahgp** [comment]
Also related to this is the fact that `color.toString()` is currently returning an unconverted rgba string. Once we figure out our direction, we can update this method to reflect it.

**Comment by lmccart** [comment]
@REAS @benfry wondering if you had this discussion for Processing and could provide any thoughts behind the behavior you've chosen in this case? I believe right now the behavior we support matches Processing (colors set before color mode retain the original numerical values).

**Comment by kadamwhite** [comment]
I'm very curious about the Processing precedent. I definitely think the way things work now is unintuitive, but I'm also curious how often `colorMode` would realistically change over the lifespan of a sketch -- how often would this be set on-the-fly as opposed to up-front?

@sarahgp As far as `toString` goes though I think 255-scale `rgba` is the best option, as I see `toString` as a method that exposes the color in a browser-compatible, normalized way -- RGBA 255 is a valid CSS string (HSB, for example, would not be) and encapsulates all the data I'd need to use the color in a CSS rule or raw canvas call, so it's (imo) the most reasonable export format. Switching how we output the string based on colormode would break interoperability with other web tools, so I would advise against it.

**Comment by sarahgp** [comment]
@kadamwhite: But hsla *is* valid. And getting an rgba string out when you specified hsla in ... would be weird.

**Comment by kadamwhite** [comment]
@sarahgp Do we support specifying HSLA? I thought we just had HSBA/RGBA

**Comment by sarahgp** [comment]
I am implementing it at this very second. PR will be next week at the latest, and besides, imho, we should be discussing things like this in terms of future expectations anyways, I think; to be sure we know  how & where we will grow.

**Comment by kadamwhite** [comment]
@sarahgp Cool, thanks, I was lacking that context. +1 for supporting HSL as a first-class mode, the web will be happier for it :)

I still think there's some value in a consistent output format from `toString` and I feel RGBA has more prevalence than HSLA across the web in general, but I don't currently have a compelling argument to support, that just intuition.

**Comment by REAS** [comment]
To translate the original example that is bringing this question to the front, we get this:

```
colorMode(RGB, 255);
color c = color(127, 255, 0);

colorMode(RGB, 1);
float myColor = red(c);
print(myColor); 
```

In this example, the print() returns 0.5. So, as far as I can imagine right now, we don't have the issue in Processing that has come up in p5.js. In Processing a "color" is a primitive data type that's an integer behind the scenes, so you never grab the "red" value out in the same way.

**Comment by brysonian** [comment]
I vaguely remember bringing up something similar to @workergnome's option 4 above a while back, looking to something like [chroma.js](https://github.com/gka/chroma.js/blob/master/doc/api.md) for inspiration. I still think there is some merit to it. Part of me feels like a new coder might be confused by `var b = rgb(0,0,255)` vs `var b = color(0,0,255)` because the later says color right there in the code. However that same coder already has to know about rgb to get anything done, so i'm not sure that really matters.

In the immediate term, could toString return something more complex, possibly even a json formatted object with all valid color space strings, and then we add toRGB, toHSL, etc methods to get more specific?

**Comment by kadamwhite** [comment]
Just opened #723 to break out the toString discussion, because it feels tangential to the purpose of this issue

**Comment by benfry** [comment]
There are two primary things at work for how the API was designed:

First, that *all* functions that impact state are forward-facing. Just like `fill()` and `stroke()`, the `colorMode()` function controls things that happen after it. All of the `xxxxMode()` functions perform this way. It's important to consider the whole API, not obsess about just a single function.

Second, `colorMode()` is truly intended to only be used once per sketch. Or perhaps there's one function that switches to a different mode, but it's primarily there for people who want to use `colorMode(RGB, 1)` (to use GL/computer graphics colors) vs the default (to 255, since people are much more used to that nowadays) vs HSB (to define color in a less crummy way than RGB).

I'm probably forgetting others, but those are the larger issues.

**Comment by workergnome** [comment]
Here's my new suggestion:

Let's have the `p5.red()`, `p5.brightness()`, et al respect `colorMode()`.  That matches the processing API, which is good for consistency.  It also makes sense, because they're global functions, and colorMode sets a global state.

For `var c = P5.color(...):

* Lets let `c.red()`, `c.green()`, and `c.blue()` return the value between [0...255].  let's let `c.red(1)` return the value between [0.0...1.0].  
* Let's let `c.alpha()` by default return [0.0...1.0], and `c.alpha(255)` return between [0...255]
* Let's do the same with HSL, with the default values being the default values from CSS.  (360, 100, 100).
* Not sure for HSB—@sarahgp, input?

And let's document that these functions explicitly return, by default, the values from CSS.  

Finally, let's extend the p5.Color constructor to take an optional [HSB | HSV | RGB] value as an optional parameter, as well as a range.  If this is not provided, use the current `colorMode`.


If I haven't missed anything, this means that by default, if you're using the library as if it were processing, with a single `colorMode()` call at the beginning, the global  color methods, and the standard color constructor, everything works out of the box.  

If you're interested in getting web colors, you can use the color object method which is where someone used to javascript would expect them to be, and by default you'll get you'll get what you expect.

If you're a power user or you want explicitly ranged values, you can manually specify what you want, and get it.

Suggestions?  Input?

**Comment by sarahgp** [comment]
I think that all makes sense @workergnome. For HSB, the defaults should also be (360, 100, 100), since that is how the color space is described.

I would highlight that we allow users to call methods not explicitly from their color space — you can call `.red()` or `brightness()` regardless of mode — and that will give you the color translated into that space with the given range, even if you get wacky.

Once I finish this adding HSLA PR, I can open issues that point to where changes are needed to support this proposal.

**Comment by brysonian** [comment]
I just have a couple of concerns. For people used to jQuery and many other js libs, something like `c.alpha(255)` looks a lot like something that would set the alpha to 255 since just about every major js lib uses the convention of getter when no args, setter when there are. Under the assumption that switch color range is a more advanced user feature, maybe we leave it to the user to be aware of the range they expect under such circumstances? It would also shift away from how the free function variations of red, alpha, etc all work, which is that you pass the color var `alpha(c)` to return the value. 

Returning alpha as 0-1 is a big change away from processing and the current p5 api which has always used 0-255. So we'd have to decide if it is worth such a big change.

Finally, the thought of `var c = color(360, 100, 100, 0.5, HSB, 360, 100, 100, 1)` seems so extreme and would really complicate the internals of the color function. Perhaps dedicated functions (`hsb()`, `rgb()`, etc) for each color space using default range's would be better? Any user getting that deep in should be fully able to use `map` (or math) to shift from a non-standard range if necessary.

**Comment by workergnome** [comment]
OK—two more thoughts, which contradict each other:

1).  I misspoke:  it's `c.getAlpha()`, not `c.alpha()`.  I know @sarahgp and I had talked about changing this, but that's what it is currently.  Clunky, but it means that you wouldn't the getters with being setters.

2).  I think that exposing the color range to the user is a bad idea in general, and  we should consider removing that ability.  It adds a ton of complexity, and I can't think of a lot of functionality that wouldn't be just as easy to implement with just color modes and good defaults.  *maybe* HDR ranges, but even then you'd want to extend or create your own object.  I feel that the range of each value is an intrinsic property of each color space, and that using it instead of just RGB_255 and RGB_FLOAT is causing us way* more headaches than it should.

**Comment by sarahgp** [comment]
@workergnome: Both `c.getAlpha` and `alpha(c)` return the alpha value, for what that's worth.

@brysonian: Currently you can call `colorMode(RGB)` to set a 255-base RGB and, as of #728, `colorMode(HSL)` sets the max values to the HSL-specified 360/100/100/1. So I don't think you'd ever need to write  `var c = color(360, 100, 100, 0.5, HSB, 360, 100, 100, 1)`? 

I still maintain the most elegant solution from a user experience point of view is to do what we do now:
* Set RGB-255 as the default, same as processing, no muss, no fuss.
* Set the defaults for HSB and HSL as specified by the color-space descriptions
* Allow users to write crazy modes if they want and get back what they put in
* Consider shifting in the rare cases mode shifts, as seems to be the case in Processing.

Either way, the functionality for most of this behavior exists in the way I structured the HSL addition, so it is there.

As for alpha, the Processing-standard 255 is the default, but if users are web-savvy enough to use web colors, I really don't think we should force adherence to an approach that no longer makes sense, given that it is not stored in a `char`.

And with the PR, I am off the climes of my GSOC project.

**Comment by brysonian** [comment]
@sarahgp I was responding to what i took @workergnome to be suggesting in https://github.com/processing/p5.js/issues/696#issuecomment-107273296

I'm not sure i follow on being web-savvy enough to use alpha, and thus it is no problem to shift to alpha being 0-1. My concern is that if we make the alpha default it will break a certain amount of existing code and possibly @lmccart's book? At the same time, the project is still technically < 1.0 so this type of API change is maybe still in the bounds of "it'll hurt a bit but it's better in the long run."

So it seems like this thread has gotten us 3 items:

1. the initial behavior @workergnome reported should be considered a bug, and his sample code should output 0.5
2. A question about whether or not the default range for alpha is 0-1 or 0-255
3. the toString issue which has been moved to #723 

Does that seem right? 

Really excited about HSL support!

**Comment by sarahgp** [comment]
@brysonian: I'm still nor sure what in @workergnome's comment would lead to that construction, but that's ok; I'm probably just being a derp.

As for alpha, on the web, with RGB and HSL both it is specified from 0 to 1, and the canvas specs I've read specify CSS color, so 0-1 alpha. So for anyone coming to p5js from a web background instead of  Processing background, that will be the expectation. However, since users can change it by setting the mode, as long as HSL defaults to 1, which it does, and calling `getAlpha()` in *that* mode returns a 0-1 value, which it does, I think we've spanned the gap.

I suspect in time it will come to be seen as a quirk of p5 but I am not sure avoiding being quirky is worth breaking the default now.

But that's a great summary of where we are. :smile_cat:

**Comment by lmccart** [comment]
Ok all sounds good.

I am ok with 0-1 for alpha if this makes more sense. Is this a change that someone has time to implement soon? I am submitting the final draft of the book very soon and can update this if need be before I do.

This string from above seems very long `var c = color(360, 100, 100, 0.5, HSB, MAX_H, MAX_S, MAX_B, MAX_A)`.  Is there really a case where you'd need to put in a different mode and max arguments just for one color and not set the colorMode for the entire sketch? I think the color stuff is already quite complicated I guess I'd worry about adding more complexity if it's not a common use case. I have only a very basic understanding of color so I apologize if I'm misunderstanding.

**Comment by brysonian** [comment]
No I don't think that use of color would need to be supported. This is exciting! Id love to take a crack but it would have to wait till next week.

**Comment by workergnome** [comment]
@brysonian— I think that @sarahgp has moved onto her Google Summer of Code Project and is not messing with the color API any more, so I think that you'd be fine to tackle it next week.  

My suggestion would be to write out the documentation first, or at least the API—maybe here, so we can all make sure that we're all on the same page?  

Let me know if there's anything I can do to help.

**Comment by sarahgp** [comment]
Yep, @brysonian & @workergnome, I am knee-deep in hardware now. Hopefully, this change won't be too crazy tho, since the HSL update also involved me taking out as many magic numbers as possible and instead mapping them to the `maxArrays` object. Please feel free to ping me if you have any questions about what I did, tho.

**Comment by workergnome** [comment]
@sarahgp—I am imagining you wading through a literal sea of Arduinos.

**Comment by brysonian** [comment]
So what is the thinking on this now? I was going to finally take a look, but it seems that folks might be ok with a default alpha as 0-255? I know part of this was depending on books and other teaching materials so it might be too late to make a change anyway.

**Comment by sarahgp** [comment]
Default alpha at 255: I think is unintuitive but trivial to switch. (Tho there is a bug in that being worked on in #842.)

Global functioning of `colorMode` is still TBD, I guess?

**Comment by xyfeng** [comment]
@lmccart @workergnome , I just created a pull request with all the fixes, please take a look at,  and let me know if there are things you want to change. #857

**Comment by lmccart** [comment]
closed with #857
